group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -474,6 +474,8 @@ public CoordinatorResult<OffsetCommitResponseData, Record> commitOffset(
         final List<Record> records = new ArrayList<>();
         final long currentTimeMs = time.milliseconds();
         final OptionalLong expireTimestampMs = expireTimestampMs(request.retentionTimeMs(), currentTimeMs);
+        groupMetadataManager.maybeUpgradeEmptyGroup(group.groupId(), records, true);
+        final int initialRecordsSize = records.size();
```

Comment:
Why are we doing this here? I am not sure to follow.

Reply:
This is because there is a `if (!records.isEmpty())` in L513. If any new records are added we need ` if (records.size() > initialRecordsSize)` instead.

Reply:
My question was more about the `maybeUpgradeEmptyGroup` called here. Why do we need it? This seems to be orthogonal to the offset metadata manager.